src/main/java/slimeknights/tconstruct/common/data/CastingRecipeBuilder.java

```diff
@@ -0,0 +1,189 @@
+import net.minecraft.advancements.ICriterionInstance;
+import net.minecraft.advancements.IRequirementsStrategy;
+import net.minecraft.advancements.criterion.RecipeUnlockedTrigger;
+import net.minecraft.data.IFinishedRecipe;
+import net.minecraft.item.Item;
+import net.minecraft.item.crafting.IRecipeSerializer;
+import net.minecraft.item.crafting.Ingredient;
+import net.minecraft.tags.Tag;
+import net.minecraft.util.IItemProvider;
+import net.minecraft.util.ResourceLocation;
+import net.minecraftforge.fluids.FluidStack;
+import net.minecraftforge.registries.ForgeRegistries;
+import slimeknights.tconstruct.library.recipe.RecipeUtil;
+import slimeknights.tconstruct.smeltery.TinkerSmeltery;
+import slimeknights.tconstruct.smeltery.recipe.AbstractCastingRecipe;
+
+import javax.annotation.Nullable;
+import java.util.function.Consumer;
+
+public class CastingRecipeBuilder {
+  private final Item result;
+  private boolean castConsumed = false;
+  private boolean switchSlots = false;
+  private FluidStack fluidStack = FluidStack.EMPTY;
+  private Ingredient cast = Ingredient.EMPTY;
+  private int coolingTime = 0;
+  private final Advancement.Builder advancementBuilder = Advancement.Builder.builder();
+  private String group;
+  private final AbstractCastingRecipe.Serializer<?> recipeSerializer;
+
+  private CastingRecipeBuilder(IItemProvider resultIn, AbstractCastingRecipe.Serializer<?> serializer) {
+    this.result = resultIn.asItem();
+    this.recipeSerializer = serializer;
+  }
+
+  public static CastingRecipeBuilder castingRecipe(IItemProvider resultIn, AbstractCastingRecipe.Serializer<?> serializer) {
+    return new CastingRecipeBuilder(resultIn, serializer);
+  }
+
+  public static CastingRecipeBuilder basinRecipe(IItemProvider resultIn) {
+    return castingRecipe(resultIn, TinkerSmeltery.basinRecipeSerializer.get());
+  }
+
+  public static CastingRecipeBuilder tableRecipe(IItemProvider resultIn) {
+    return castingRecipe(resultIn, TinkerSmeltery.tableRecipeSerializer.get());
+  }
+
+  public CastingRecipeBuilder addCriterion(String name, ICriterionInstance criterionInstance) {
+    this.advancementBuilder.withCriterion(name, criterionInstance);
+    return this;
+  }
+
+  public CastingRecipeBuilder setFluid(FluidStack fluidStack) {
+    this.fluidStack = fluidStack;
+    return this;
+  }
+  public CastingRecipeBuilder setGroup(String groupIn) {
+    this.group = groupIn;
+    return this;
+  }
+
+  public CastingRecipeBuilder setCast(Tag<Item> tagIn, boolean consumed) {
+    return this.setCast(Ingredient.fromTag(tagIn), consumed);
+  }
+
+  public CastingRecipeBuilder setCast(IItemProvider itemIn, boolean consumed) {
+    return this.setCast(Ingredient.fromItems(itemIn), consumed);
+  }
+
+  public CastingRecipeBuilder setCast(Ingredient ingredient, boolean consumed) {
+    this.cast = ingredient;
+    this.castConsumed = consumed;
+    return this;
+  }
+
+  public CastingRecipeBuilder setSwitchSlots() {
```

Comment:
Might we worth Javadocs on a few of these more obscure methods especially. It is a bit redundant to the recipe class, but most people will use the builder.

Reply:
I javadoc'd this one. what would you also consider obscure here?

Reply:
Switch slots is a little vague, since from the outside viewpoint the casting blocks only have one slot.